modules/RtcStats/RtcStats.js

```diff
@@ -0,0 +1,138 @@
+import Logger from '@jitsi/logger';
+import rtcstatsInit from '@jitsi/rtcstats/rtcstats';
+import traceInit from '@jitsi/rtcstats/trace-ws';
+
+const logger = Logger.getLogger(__filename);
+
+/**
+ * Filter out RTCPeerConnection that are created by callstats.io.
+ *
+ * @param {*} config - Config object sent to the PC c'tor.
+ * @returns {boolean}
+ */
+function connectionFilter(config) {
+    if (config?.iceServers[0] && config.iceServers[0].urls) {
+        for (const iceUrl of config.iceServers[0].urls) {
+            if (iceUrl.indexOf('callstats.io') >= 0) {
+                return true;
+            }
+        }
+    }
+}
+
+/**
+ * Class the controls the rtcstats node module.
+ */
+class RtcStats {
+    initialized = false;
+    trace = null;
+    eventCallback = null;
+    closeCallback = null;
+    setTrace;
+    traceConfigured = new Promise(r => {
+        this.setTrace = r;
+    });
+    meetingFqn = null;
+    setMeetingF;
+    meetingFqnSet = new Promise(r => {
+        this.setMeetingF = r;
+    });
+
+    /**
+     * Initializes the underlying rtcstats node moduke.
+     */
+    init(options) {
+        const { rtcstatsUseLibJitsi } = options.analytics;
+
+        if (rtcstatsUseLibJitsi) {
+            if (!this.initialized) {
+                this.meetingFqnSet.then(() => {
```

Comment:
I'm assuming that this will somehow reset when adding the de-init step which @saghul mentioned, otherwise the promise will be called a second time on mobile (when a user joins another meeting) but by that time it will already have been resolved.

Reply:
From my testing on mobile the worries around re-init are already been taken care of by George code, which re-inits all of the rtcstats code. But I'll double check for the room name to make sure.

Reply:
The scenario which could happen (I think). Is the following.

`if (changed) {
            this.reset();

            if (newOptions.meetingFqn && newOptions.endpoint) {
                this.init(newOptions);
            } else {
                logger.warn('RTCStats is enabled but is waiting for meetingFqn and endpoint.');
            }
        }
    }
   `
   This would trigger
   `
       init(options: InitOptions) {

        const { useLibRtcStats, meetingFqn } = options;

        if (useLibRtcStats) {
            logger.info('setting up libRtcStats');

            const rtcOptions = {
                meetingName: meetingFqn,
                eventCallback: this.handleRtcstatsEvent.bind(this),
                onCloseCallback: this.handleTraceWSClose.bind(this)
            };

            JitsiMeetJS.setRtcStatsMeetingFqn(rtcOptions);

            JitsiMeetJS.getRtcStatsTrace().then((traceF: any) => {
                this.setTrace(traceF);
            });

        }
   `
So on mobile when a user joins another room, we would reset the connection (delete the trace) but `JitsiMeetJS.setRtcStatsMeetingFqn(rtcOptions);`  would be called again and by this point the promise was already resolved, additional resolve calls would have no effect, unless we somehow reset the promise.
From what I understand ljm init will be called a second time but the RtcStats instance inside is a singleton so the class and hence the promise doesn't get re-initialized, @saghul is this right?

Reply:
The scenario which could happen (I think). Is the following.
```
if (changed) {
            this.reset();

            if (newOptions.meetingFqn && newOptions.endpoint) {
                this.init(newOptions);
            } else {
                logger.warn('RTCStats is enabled but is waiting for meetingFqn and endpoint.');
            }
        }
    }
```
This would trigger:
```
     init(options: InitOptions) {

        const { useLibRtcStats, meetingFqn } = options;

        if (useLibRtcStats) {
            logger.info('setting up libRtcStats');

            const rtcOptions = {
                meetingName: meetingFqn,
                eventCallback: this.handleRtcstatsEvent.bind(this),
                onCloseCallback: this.handleTraceWSClose.bind(this)
            };

            JitsiMeetJS.setRtcStatsMeetingFqn(rtcOptions);

            JitsiMeetJS.getRtcStatsTrace().then((traceF: any) => {
                this.setTrace(traceF);
            });

        }
 ```
So on mobile when a user joins another room, we would reset the connection (delete the trace) but JitsiMeetJS.setRtcStatsMeetingFqn(rtcOptions);  would be called again and by this point the promise was already resolved, additional resolve calls would have no effect, unless we somehow reset the promise.
From what I understand ljm init will be called a second time but the RtcStats instance inside is a singleton so the class and hence the promise doesn't get re-initialized, @saghul is this right?

Reply:
Sounds about right! Which is wrong :-P